### ISSUE TITLE
Pantry panic in a4x2: Job completed while not InProgress: Done

### ISSUE THREAD

**Issue by jgallagher** [oxidecomputer/crucible#1513]
I was trying to import a disk in a4x2 via this Oxide CLI invocation:

```
oxide --timeout 120 disk import \
    --path debian-12-genericcloud-amd64.raw \
    --disk bookworm-boot \
    --project prototypes \
    --description "Debian bookworm disk" \
    --snapshot bookworm-snapshot \
    --image bookworm \
    --image-description "Debian bookworm image" \
    --image-os Debian \
    --image-version 12
```

After not quite 2 minutes it failed with

```
sending chunk to thread failed: channel closed
```

which seems similar to the test flake oxidecomputer/omicron#6771, but in this case it looks like the pantry panicked:

```
thread 'tokio-runtime-worker' panicked at upstairs/src/client.rs:1280:13:
[1] Job 1252 completed while not InProgress: Done DownstairsIO { ... trim a couple megabytes ... }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[ Oct 18 18:21:28 Stopping because all processes in service exited. ]
```

The full pantry log: [pantry.log](https://github.com/user-attachments/files/17439584/pantry.log)

**Comment by mkeeter** [comment]
This looks like a race condition in replay!

When a downstairs is `Offline`, we still send jobs to it:

https://github.com/oxidecomputer/crucible/blob/bfcc86f79dfa14b619735d4b81a005d262ad1e59/upstairs/src/client.rs#L968-L974

We actually send IO jobs to the client task, which sends them out!  However, they're ignored by the Downstairs while it's in negotiation.

Now, imagine sending a write _exactly when_ the downstairs finishes negotiation:

- Downstairs sends the final negotiation message (`Message::ExtentVersions`) and decides it's done with negotiation!
- A write is sent to the ostensibly-offline Downstairs
- The Downstairs thinks it's online, so it executes the write and sends a reply
- The Upstairs receives the final negotiation message
  - It replays all IO jobs, including the one it just sent to the Downstairs!
  - The Downstairs replies to that one as well
- The upstairs receives both replies
  - It panics because the job is completed twice

--------------------------------------------------------------------------------

Evidence for this theory is in the `JobId` numbers:

```
{"msg":"[1] Transition from Offline to Active","v":0,"name":"crucible-pantry","level":30,"time":"2024-10-18T18:20:23.406003244Z","hostname":"oxz_crucible_pantry_cea84efd-1d0a-4db6-bbbc-a65668168c5c","pid":23205,"client":"1","":"downstairs","session_id":"a34a07ad-362b-4637-999a-cdb19f1a4874","component":"datafile"}
{"msg":"[1] client re-new 243 jobs since flush 1009","v":0,"name":"crucible-pantry","level":30,"time":"2024-10-18T18:20:23.406017306Z","hostname":"oxz_crucible_pantry_cea84efd-1d0a-4db6-bbbc-a65668168c5c","pid":23205,"":"downstairs","":"downstairs","session_id":"a34a07ad-362b-4637-999a-cdb19f1a4874","component":"datafile"}
{"msg":"[1] Replayed 243 jobs since flush: 1009","v":0,"name":"crucible-pantry","level":30,"time":"2024-10-18T18:20:23.408376287Z","hostname":"oxz_crucible_pantry_cea84efd-1d0a-4db6-bbbc-a65668168c5c","pid":23205,"":"downstairs","":"downstairs","session_id":"a34a07ad-362b-4637-999a-cdb19f1a4874","component":"datafile"}
```

The upstairs replays 243 jobs, starting at `JobId(1009)`.  In other words, the last job replayed is `JobId(1252)` (the flush itself isn't replayed).

It's this final job that fails:

```
[1] Job 1252 completed while not InProgress: Done DownstairsIO { ds_id: JobId(1252), guest_id: GuestWorkId(253), work: Write { dependencies: [JobId(1251)], start_eid: ExtentId(7), start_offset: BlockOffset(26624), blocks: [elided], data: b"elided" }, state: ClientData([InProgress, Done, Done]), acked: true, replay: true, data: None, read_validations: [], backpressure_guard: ClientMap(ClientData([Some(BackpressureGuard { counter: BackpressureCountersInner { write_bytes: 57671680, jobs: 180, io_bytes: 57671680 }, write_bytes: 524288, io_bytes: 524288 }), None, None])) }
```

--------------------------------------------------------------------------------

I'm going to try to reproduce this in unit tests, then probably drop jobs if the downstairs is offline.

**Comment by mkeeter** [comment]
Should be fixed in #1519
